flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1505 @@
+		int keyLen = keyBytes.length;
+
+		return removeNode(keyByteBuffer, 0, keyLen, true);
+	}
+
+	@Override
+	public <T> void transform(
+		K key,
+		N namespace,
+		T value,
+		StateTransformationFunction<S, T> transformation) throws Exception {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		long node = getNode(keyByteBuffer, 0, keyLen);
+		S oldState = node == NIL_NODE ? null : helpGetNodeState(node);
+		S newState = transformation.apply(oldState, value);
+		byte[] stateBytes = skipListValueSerializer.serialize(newState);
+		putNode(keyByteBuffer, 0, keyLen, stateBytes, false);
+	}
+
+	// Detail implementation methods ---------------------------------------------------------------
+
+	/**
+	 * Find the node containing the given key.
+	 *
+	 * @param keyByteBuffer byte buffer storing the key.
+	 * @param keyOffset     offset of the key.
+	 * @param keyLen        length of the key.
+	 * @return id of the node. NIL_NODE will be returned if key does no exist.
+	 */
+	private long getNode(ByteBuffer keyByteBuffer, int keyOffset, int keyLen) {
+		int deleteCount = 0;
+		long prevNode = findPredecessor(keyByteBuffer, keyOffset, 1);
+		long currentNode  = helpGetNextNode(prevNode, 0);
+		long nextNode;
+
+		int c;
+		while (currentNode != NIL_NODE) {
+			nextNode = helpGetNextNode(currentNode, 0);
+
+			boolean isRemoved = isNodeRemoved(currentNode);
+			if (isRemoved) {
+				// remove the node physically when there is no snapshot running
+				if (highestRequiredSnapshotVersion == 0 && deleteCount < numKeysToDeleteOneTime) {
+					doPhysicalRemove(currentNode, prevNode, nextNode);
+					logicallyRemovedNodes.remove(currentNode);
+					totalSize--;
+					deleteCount++;
+				} else {
+					prevNode = currentNode;
+				}
+				currentNode = nextNode;
+				continue;
+			}
+
+			c = compareByteBufferAndNode(keyByteBuffer, keyOffset, keyLen, currentNode);
+
+			// find the key
+			if (c == 0) {
+				return currentNode;
+			}
+
+			// the key is less than the current node, and nodes after current
+			// node can not be equal to the key.
+			if (c < 0) {
+				break;
+			}
+
+			prevNode = currentNode;
+			currentNode = helpGetNextNode(currentNode, 0);
+		}
+
+		return NIL_NODE;
+	}
+
+	/**
+	 * Put the key into the skip list. If the key does not exist before, a new node
+	 * will be created. If the key exists before, the old state will be returned.
+	 *
+	 * @param keyByteBuffer  byte buffer storing the key.
+	 * @param keyOffset      offset of the key.
+	 * @param keyLen         length of the key.
+	 * @param value          the value.
+	 * @param returnOldState whether to return old state.
+	 * @return the old state. Null will be returned if key does not exist or returnOldState is false.
+	 */
+	private S putNode(ByteBuffer keyByteBuffer, int keyOffset, int keyLen, byte[] value, boolean returnOldState) {
```

Comment:
`keyOffset` seems to be always `0`.

Reply:
will add a UT case to check against the non-zero offset correctness.